api/v1alpha1/linodemachine_webhook.go

```diff
@@ -0,0 +1,196 @@
+	var (
+		errs   field.ErrorList
+		client = linodego.NewClient(http.DefaultClient)
+		ctx    = context.TODO()
+	)
+	if err := validateRegion(ctx, client, r.Spec.Region, field.NewPath("spec").Child("region")); err != nil {
+		errs = append(errs, err)
+	}
+	plan, err := validateLinodeType(ctx, client, r.Spec.Type, field.NewPath("spec").Child("type"))
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if err := r.validateLinodeMachineDisks(plan); err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+func validateRegion(ctx context.Context, client linodego.Client, region string, path *field.Path) *field.Error {
+	if _, err := client.GetRegion(ctx, region); err != nil {
+		return field.NotFound(path, region)
+	}
+
+	return nil
+}
+
+func validateLinodeType(ctx context.Context, client linodego.Client, plan string, path *field.Path) (*linodego.LinodeType, *field.Error) {
+	linodeType, err := client.GetType(ctx, plan)
+	if err != nil {
+		return nil, field.NotFound(path, plan)
+	}
+
+	return linodeType, nil
+}
+
+func (r *LinodeMachine) validateLinodeMachineDisks(plan *linodego.LinodeType) *field.Error {
+	// The Linode plan information is required to perform disk validation
+	if plan == nil {
+		return nil
+	}
+
+	var (
+		// The Linode API represents storage sizes in megabytes (MB)
+		// See: https://www.linode.com/docs/api/linode-types/#type-view
+		planSize   = resource.MustParse(fmt.Sprintf("%d%s", plan.Disk, "M"))
+		remainSize = &planSize
+		err        *field.Error
+	)
+
+	if remainSize, err = validateDisk(r.Spec.OSDisk, field.NewPath("spec").Child("osDisk"), remainSize); err != nil {
+		return err
+	}
+	if _, err := validateDataDisks(r.Spec.DataDisks, field.NewPath("spec").Child("dataDisks"), remainSize); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateDataDisks(disks map[string]*InstanceDisk, path *field.Path, availSize *resource.Quantity) (*resource.Quantity, *field.Error) {
+	var (
+		devs       []string
+		remainSize = availSize
+	)
+
+	for dev, disk := range disks {
+		if !slices.Contains(LinodeMachineSCSIPaths, dev) {
+			return nil, field.Forbidden(path.Child(dev), fmt.Sprintf("allowed device paths: %v", LinodeMachineSCSIPaths))
+		}
+		if slices.Contains(devs, dev) {
+			return nil, field.Duplicate(path.Child(dev), "duplicate device path")
+		}
+		devs = append(devs, dev)
+
+		var err *field.Error
+		if remainSize, err = validateDisk(disk, path.Child(dev), remainSize); err != nil {
+			return nil, err
+		}
+	}
+	return remainSize, nil
+}
+
+func validateDisk(disk *InstanceDisk, path *field.Path, availSize *resource.Quantity) (*resource.Quantity, *field.Error) {
+	if disk == nil {
+		return availSize, nil
+	}
+
+	if disk.Size.Sign() == -1 {
+		return nil, field.Invalid(path, disk.Size.String(), "cannot be negative")
+	}
+	if availSize.Cmp(disk.Size) == -1 {
+		return nil, field.Invalid(path, disk.Size.String(), "sum of disk sizes cannot exceed plan storage")
```

Comment:
we should probably also say what the plan storage available is 
```suggestion
		return nil, field.Invalid(path, disk.Size.String(), fmt.Sprintf("sum of disk sizes cannot exceed plan storage of %s", availSize.Size.String()))
```

Reply:
FWIW, the `availSize` in this helper function is the amount of the _remaining_ storage (i.e. you decrement this value every time you process a disk).

But yeah, you got me 🙈 I didn't want to pass around non-function arguments cause it's un-aesthetic ✨✨